pandas/core/nanops.py

```diff
@@ -254,9 +254,16 @@ def nansum(values, axis=None, skipna=True):
 @bottleneck_switch()
 def nanmean(values, axis=None, skipna=True):
     values, mask, dtype, dtype_max = _get_values(values, skipna, 0)
-    the_sum = _ensure_numeric(values.sum(axis, dtype=dtype_max))
     count = _get_counts(mask, axis)
 
+    dtype_sum = dtype_max
+    if is_integer_dtype(dtype):
+        dtype_sum = np.float64
+    elif is_float_dtype(dtype):
+        dtype_sum = dtype
+        count = dtype.type(count)
```

Comment:
hmm, ok. 

maybe better to add pass thru the dtype to ``get_counts`` then as well (and cast to that inside ``get_counts``), to avoid this specific code all over the place